### PR TITLE
Use CURLOPT_CONNECTTIMEOUT if timeout too big

### PR DESCRIPTION
**Pull request by LorenzoLeonardo**
If the time that was converted into millisecond
is more than the capacity of c_long, it will
overflow, thus we need to convert it into seconds
and use CURLOPT_CONNECTTIMEOUT.

### CODE REVIEW

File: src/easy/handler.rs [hunk 1]
```diff
@@ -2049,7 +2050,14 @@ impl<H> Easy2<H> {
     /// `CURLOPT_CONNECTTIMEOUT_MS`.
     pub fn connect_timeout(&mut self, timeout: Duration) -> Result<(), Error> {
         let ms = timeout.as_secs() * 1000 + timeout.subsec_millis() as u64;
```
Comment (alexcrichton) [inline src/easy/handler.rs:2052]: While not 100% related to this change, while you're here would you be up for using [`Duration::as_millis`](https://doc.rust-lang.org/stable/std/time/struct.Duration.html#method.as_millis) here instead? The fallible u128-to-`c_long` conversion below will automatically handle out-of-range values of `Duration` itself.

File: src/easy/handler.rs [hunk 2]
```diff
@@ -2049,7 +2050,14 @@ impl<H> Easy2<H> {
...
-        self.setopt_long(curl_sys::CURLOPT_CONNECTTIMEOUT_MS, ms as c_long)
+
+        match c_long::try_from(ms) {
+            Ok(amt) => self.setopt_long(curl_sys::CURLOPT_CONNECTTIMEOUT_MS, amt),
+            Err(_) => {
+                let amt = c_long::try_from(ms / 1000).unwrap();
```
Comment (alexcrichton) [inline src/easy/handler.rs:2057]: On second thought (after commenting on https://github.com/alexcrichton/curl-rust/pull/495), could this `.unwrap()` be translated to an error to be returned here instead? That way it's possible for callers to be notified that timeouts are simply too large.

Comment (LorenzoLeonardo) [reply src/easy/handler.rs:2057]: I tried to search the best error code to handle this for curl, but I cannot find that is related to overflow errors.

Comment (alexcrichton) [reply src/easy/handler.rs:2057]: Perhaps `CURLE_BAD_FUNCTION_ARGUMENT`?

Comment (LorenzoLeonardo) [reply src/easy/handler.rs]: Please check, I added a Unit Test too

**Review by alexcrichton — APPROVED** [review]
I think tests may be failing though? (perhaps because curl is too old in the container?)

**Comment by LorenzoLeonardo** [comment]
> I think tests may be failing though? (perhaps because curl is too old in the container?)

I will try to use i32:MAX instead of std::i32:MAX
https://doc.rust-lang.org/std/i32/constant.MAX.html

**Comment by LorenzoLeonardo** [comment]
oh, that's weird, if I run the Unit Test locally this will succeed.

**Comment by alexcrichton** [comment]
I believe the failure is because `c_ulong` is u64 on Linux so `u64::MAX` fits. I think you'll want to specify a larger amount than `u64::MAX` milliseconds to guarantee failure.

**Comment by LorenzoLeonardo** [comment]
will try to use from_secs() to guarantee max value

**Comment by LorenzoLeonardo** [comment]
It seems there are test error not related to the change.

**Comment by LorenzoLeonardo** [comment]
it is ok now

**Pull request merged**